src/CoreWCF.Primitives/src/CoreWCF/Dispatcher/SecurityServiceDispatcher.cs

```diff
@@ -0,0 +1,565 @@
+        //    }
+        //    this.identity = result;
+        //}
+
+        private void ThrowIfProtocolFactoryNotSet()
+        {
+            if (this.securityProtocolFactory == null)
+            {
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.SecurityProtocolFactoryShouldBeSetBeforeThisOperation)));
+            }
+        }
+
+        public async Task<IServiceChannelDispatcher> CreateServiceChannelDispatcherAsync(IChannel outerChannel)
+        { 
+            //TODO should have better logic
+            //Initialization path start
+            if (outerChannel.ChannelDispatcher == null)
+            {
+                TypedChannelDemuxer typedChannelDemuxer = this.ChannelBuilder.ChannelDemuxer.GetTypedServiceDispatcher<IReplyChannel>();
+                IServiceChannelDispatcher channelDispatcher = await typedChannelDemuxer.CreateServiceChannelDispatcherAsync(outerChannel);
+                return channelDispatcher;
+            }
+            //initialization end
+            //Below dispatches all SCT call for first time for all clients
+            else
+            {
+                IServiceChannelDispatcher securityReplyChannelDispatcher = GetInnerChannelDispatcher(outerChannel);
+                return securityReplyChannelDispatcher;
+            }
+        }
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="outerChannel"></param>
+        /// <returns></returns>
+        internal IServiceChannelDispatcher GetAuthChannelDispatcher(IReplyChannel outerChannel)
+        {
+            if (securityAuthServiceChannelDispatcher == null)
+            {
+                lock (ThisLock)
+                {
+                    if (securityAuthServiceChannelDispatcher == null)
+                    {
+                        Task<IServiceChannelDispatcher> channelTask = SecurityAuthServiceDispatcher.CreateServiceChannelDispatcherAsync(outerChannel);
+                        securityAuthServiceChannelDispatcher = channelTask.GetAwaiter().GetResult();
+                    }
+                }
+            }
+            return securityAuthServiceChannelDispatcher;
+        }
+
+
+        /// <summary>
+        /// Return same instance of real service channel dispatcher.
+        /// This method is called by SecurityReplySessionServiceChannelDispatcher(SecuritySessionServerSettings) to dispatch real service 
+        /// </summary>
+        /// <param name="outerChannel"></param>
+        /// <returns></returns>
+        internal IServiceChannelDispatcher GetInnerServiceChannelDispatcher(IReplyChannel outerChannel)
+        {
+            lock (ThisLock)
+            {
+                Task<IServiceChannelDispatcher> channelTask = InnerServiceDispatcher.CreateServiceChannelDispatcherAsync(outerChannel);
+                return channelTask.GetAwaiter().GetResult();
+            }
```

Comment:
I don't like sync over async code. Something needs to be reworked here. There is no saved state in this class related to the returned dispatcher so the concurrency issue isn't in this class. If there's a problem with the inner service dispatcher not being thread safe calling CreateServiceChannelDispatcherAsync, then this won't fix it. You could have multiple endpoints using the same inner service channel dispatcher and the thread safety problem returns. The fix needs to be in the InnerServiceDispatcher.CreateServiceChannelDispatcherAsync implementation which would allow this method to be async.